### DISCUSSION TITLE
error with "Claim free JCB package key"

### DISCUSSION THREAD

**Discussion by linpi1** [joomengine/Joomla-Component-Builder#804] [General]
I starred the Joomla Member Manager, and on https://www.joomlacomponentbuilder.com/claim-free-package-key I always get the error in red by the gitHub username "Please complete all required fields!"  

But I put the real github username (linpi1) the email address, the Repository & the Relationship. What else can I do ?

**Comment by Llewellynvdm** [comment]
Seems like we have a problem in the form, let me take a quick look and see if I can get it resolved.

**Comment by counterpoint** [comment]
Same for me.

On 14/09/2021 21:47, Pierre wrote:
> I starred the Joomla Member Manager, and on 
> https://www.joomlacomponentbuilder.com/claim-free-package-key 
> <https://www.joomlacomponentbuilder.com/claim-free-package-key> I always 
> get the error in red by the gitHub username "Please complete all 
> required fields!"
> 
> But I put the real github username (linpi1) the email address, the 
> Repository & the Relationship. What else can I do ?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/vdm-io/Joomla-Component-Builder/discussions/804>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAJG5RNIVAYGZPXXWCZ6ESTUB6YE3ANCNFSM5EA7CC6A>.
> Triage notifications on the go with GitHub Mobile for iOS 
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> 
> or Android 
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> 
> 

-- 
Martin Brampton, Black Sheep Research
***@***.***
https://black-sheep-research.com
Tel: +44 1653 696210
Mobile: +44 7771 787010
Office: 4 York Road, Malton, North Yorkshire, YO17 6AX, UK.

**Comment by Llewellynvdm** [comment]
Seems like RSForm has changed their events API and so [my plugin](https://www.vdm.io/joomla/item/rsform-advance-php-plugin) that target their events are no longer working.... :-1: 

The whole idea of the form was to encourage everyone to interact with the GitHub repositories, and for their trouble I give them the keys to the JCB packages. At first I did it manually and then I moved the validation to this RSForm... but now it seems like a far bigger task then tweaking two lines.

**Reply by linpi1** [reply]
ok I understand. It would great to get this key to be able to install it JCE.

**Comment by Llewellynvdm** [comment]
Okay I have [fixed the form](https://www.joomlacomponentbuilder.com/claim-free-package-key), please try again :+1:

**Reply by linpi1** [reply]
Yes wonderful, it works and I could already install the Joomla Member Manager in JCB. Thanks a lot! :-)

**Comment by counterpoint** [comment]
I still have the same problem, although I have watched Joomla Component Builder, my key request still gets "Please complete all required fields!"

**Reply by Llewellynvdm** [reply]
Wow I thought that was resolved... let me see, what browser are you using, please try Firefox.

**Reply by Llewellynvdm** [reply]
I just tested it again:

![image](https://user-images.githubusercontent.com/5607939/134933818-11ac46d6-28d0-4bc6-bfab-ef24272ac762.png)
> Obviously use correct data...

![image](https://user-images.githubusercontent.com/5607939/134933969-aa4536d6-50da-4e52-bdb1-3efa0a65d95b.png)

> And it gave me the key no problem.

**Comment by counterpoint** [comment]
I am using Firefox. Entering "counterpoint" and "xxxxxx@xxxxxxxx.com". Selecting "Joomla Component Builder" and "watched". Plus the reCaptcha. Attaching screenshot of submission rejected and showing that I am logged in as "counterpoint" and am watching JCB (note the "unwatch" button).
![Screenshot from 2021-09-28 10-35-03](https://user-images.githubusercontent.com/1208005/135062845-5b5af3e3-e5bb-4aa6-b8d1-a45f793d6ac3.png)
![Screenshot from 2021-09-28 10-35-59](https://user-images.githubusercontent.com/1208005/135062871-98948096-4f46-4825-bcaa-0803d3d8a450.png)

**Reply by Llewellynvdm** [reply]
Okay it does seem like GitHub only lists you as a watcher when you have selected "All Activity":
![image](https://user-images.githubusercontent.com/5607939/135069131-3e6f0710-0ce2-4a26-a781-30b3982d6170.png)

I know that could be painful... the many notices one tend to get... so try staring or forking.

**Reply by counterpoint** [reply]
Thanks, I already had all activity ticked all the time. Still didn't work. Something seems to be broken there. Added starring. That does seem to work. Got a key.

**Reply by Llewellynvdm** [reply]
I just tested it with your username.... and it worked for me. So this behavior is very strange... we are just calling the API:

https://api.github.com/repos/vdm-io/Joomla-Component-Builder/watchers?per_page=100&page=4

### I am really sorry for the inconvenience caused...

We really just checking the login value to match your username like this:

```php
function checkString($string)
{
	if (isset($string) && is_string($string) && strlen($string) > 0)
	{
		return true;
	}
	return false;
}

function checkJson($string)
{
	if (checkString($string))
	{
		json_decode($string);
		return (json_last_error() === JSON_ERROR_NONE);
	}
	return false;
}

function checkArray($array, $removeEmptyString = false)
{
	if (isset($array) && is_array($array) && count($array) > 0)
	{
		// also make sure the empty strings are removed
		if ($removeEmptyString)
		{
			foreach ($array as $key => $string)
			{
				if (empty($string))
				{
					unset($array[$key]);
				}
			}
			return checkArray($array, false);
		}
		return true;
	}
	return false;
}

function getFileContents($path, $none = '')
{
	if (checkString($path))
	{
		// use curl if available
		if (function_exists('curl_version'))
		{
			// start curl
			$ch = curl_init();
			// set the options
			$options = array();
			$options[CURLOPT_URL] = $path;
			$options[CURLOPT_USERAGENT] = 'Mozilla/5.0 (Windows; U; Windows NT 6.1; en-US; rv:1.9.2.12) Gecko/20101026 Firefox/3.6.12';
			$options[CURLOPT_RETURNTRANSFER] = TRUE;
			$options[CURLOPT_SSL_VERIFYPEER] = FALSE;
			$options[CURLOPT_HTTPHEADER] = ['Authorization: token xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'];
			// load the options
			curl_setopt_array($ch, $options);
			// get the content
			$content = curl_exec($ch);
			// close the connection
			curl_close($ch);
			// return if found
			if (checkString($content))
			{
				return $content;
			}
		}
	}
	return $none;
}

function getValues($get, $user, $page, &$search)
{
	$repos = getFileContents($get.$page);
	if (checkJson($repos))
	{
		$repos = json_decode($repos, true);
		// check if this is an error
		if (!isset($repos['message']) && checkArray($repos))
		{
			foreach ($repos as $repo)
			{
				if (isset($repo['owner']) && isset($repo['owner']['login']) & strtolower($repo['owner']['login']) == $user)
				{
					return true;
				}
				elseif (isset($repo['login']) && strtolower($repo['login']) == $user)
				{
					return true;
				}
			}
			return false;
		}
	}
	$search = false;
	return $search;
}

function checkRelationship($user, $type, $project)
{
	// set some validation
	$types = array('forks' => 'forks', 'stargazers' => 'stargazers', 'watchers' => 'watchers');
	$found = false;
	$projects = array(
		'jcb' => 'vdm-io/Joomla-Component-Builder',
		'jmm' => 'Llewellynvdm/Joomla-Members-Manager',
		'sd' => 'Llewellynvdm/Joomla-Sermon-Distributor',
		'qna' => 'Llewellynvdm/Joomla-Questions-and-Answers',
		'demo' => 'Llewellynvdm/Joomla-Demo-Component',
		'sg' => 'Llewellynvdm/Joomla-Support-Groups',
		'cbp' => 'Llewellynvdm/Joomla-Cost-Benefit-Projection',
		'ehp' => 'Namibia/eHealth-Portal');

	if (isset($types[$type]) && isset($projects[$project]))
	{
		// set urls
		$get = 'https://api.github.com/repos/'.$projects[$project].'/'.$type.'?per_page=100&page=';
		$page = 1;
		$search = true;
		// search
		while ($search && !$found)
		{
			$found = getValues($get, $user, $page, $search);
			$page++;
		}
	}
	return $found;
}

if (isset($args['post']) && isset($args['post']['github_name']) && checkString($args['post']['github_name']) &&
	isset($args['post']['rel']) && checkArray($args['post']['rel']) && checkString($args['post']['rel'][0]) &&
	isset($args['post']['project']) && checkArray($args['post']['project']) && checkString($args['post']['project'][0])
)
{
	$userName = preg_replace("([^\w\s\d\-_\(\)])", '', $args['post']['github_name']);
	$supportedUs = checkRelationship(strtolower(trim($userName)), strtolower(trim($args['post']['rel'][0])), strtolower(trim($args['post']['project'][0])));
}

if (!$supportedUs)
{
	$args['invalid'][] = 85;
}
```
Any help improving this is welcome :+1:
